File: CPP/7_.cpp
#include <iostream>
#include <cmath>
#include <vector>
#include <cstdlib>
#include <ctime>

using namespace std;

int gcd(int a, int b)
{
    if (a < b)
    {
        return gcd(b, a);
    }
    else if (a % b == 0)
    {
        return b;
    }
    else
    {
        return gcd(b, a % b);
    }
}

int power(int a, int b, int c)
{
    int x = 1;
    int y = a;

    while (b > 0)
    {
        if (b % 2 != 0)
        {
            x = (x * y) % c;
        }
        y = (y * y) % c;
        b = b / 2;
    }

    return x % c;
}

int gen_key(int q)
{
    long long int key = rand() % (int)pow(10, 10) + q;
    while (gcd(q, key) != 1)
    {
        key = rand() % (int)pow(10, 10) + q;
    }
    return key;
}

vector<int> encrypt(const string &msg, int q, int h, int g)
{
    vector<int> en_msg(msg.length());

    int k = gen_key(q);
    int s = power(h, k, q);
    int p = power(g, k, q);

    cout << "g^k used : " << p << endl;
    cout << "g^ak used : " << s << endl;

    for (size_t i = 0; i < msg.length(); ++i)
    {
        en_msg[i] = s * int(msg[i]);
    }

    cout << "en_msg : ";
    for (int val : en_msg)
    {
        cout << val << " ";
    }
    cout << endl;

    return en_msg;
}

string decrypt(const vector<int> &en_msg, int p, int key, int q)
{
    string dr_msg;

    int h = power(p, key, q);
    for (int val : en_msg)
    {
        dr_msg.push_back(char(val / h));
    }

    return dr_msg;
}

int main()
{
    srand(time(0));

    string msg = "Ajinkya";
    cout << "Original Message: " << msg << endl;

    long long int q = rand() % (int)pow(10, 20) + pow(10, 10);
    long long int g = rand() % (q - 2) + 2;

    cout << "q used : " << q << endl;
    cout << "g used : " << g << endl;

    int key = gen_key(q);
    cout << "Private Key : " << key << endl;

    int h = power(g, key, q);
    cout << "g^a used : " << h << endl;

    vector<int> en_msg = encrypt(msg, q, h, g);
    string dr_msg = decrypt(en_msg, h, key, q);

    cout << "Decrypted Message : " << dr_msg << endl;

    return 0;
}

File: Python/7_.py
# Python program to illustrate ElGamal encryption

# Very Large Number q
# q = Number between 10^20 and 10^50
# g = Number between 2 and q
# key = Number between 10^20 and q which is coprime with q
# h = g^key mod q



import random, math


def gcd(a, b):
    if a < b:
        return gcd(b, a)
    elif a % b == 0:
        return b
    else:
        return gcd(b, a % b)
	

def power(a, b, c):
    x = 1
    y = a

    while b > 0:
        if b % 2 != 0:
            x = (x * y) % c
        y = (y * y) % c
        b = int(b / 2)

    return x % c


def gen_key(q):

	key = random.randint(pow(10, 20), q)
	while math.gcd(q, key) != 1:
		key = random.randint(pow(10, 20), q)
	return key

def encrypt(msg, q, h, g):

	en_msg = list(msg)

	k = gen_key(q)  # Private key for sender
	s = pow(h, k, q)
	p = pow(g, k, q)

	print("g^k used : ", p)
	print("g^ak used : ", s)

	for i in range(0, len(en_msg)):
		en_msg[i] = s * ord(en_msg[i])

	print("en_msg : ", en_msg)

	return en_msg, p


def decrypt(en_msg, p, key, q):

	dr_msg = []
	h = pow(p, key, q)
	for i in range(0, len(en_msg)):
		dr_msg.append(chr(int(en_msg[i]/h)))

	return dr_msg


def main():

	msg = 'Ajinkya'
	print("Original Message :", msg)

	q = random.randint(pow(10, 20), pow(10, 50))
	g = random.randint(2, q)
	
	print("q used : ", q)
	print("g used : ", g)

	key = gen_key(q)  
	print("Private Key :", key)

	h = pow(g, key, q)

	print("g^a used : ", h)

	en_msg, p = encrypt(msg, q, h, g)
	dr_msg = decrypt(en_msg, p, key, q)
	dmsg = ''.join(dr_msg)
	print("Decrypted Message :", dmsg)


if __name__ == '__main__':
	main()
